그래프 차트 화면 구현

donggeonoh opened the pull request:
- 그래프 뷰 레이아웃 추가
- Candlestick API 추가 및 데이터 불러오는 기능 구현
- WebSocket API 추가 및 데이터 갱신하는 기능 구현

 #50

donggeonoh committed:
feat: Chart SPM 추가 및 코인 차트 뷰 추가 및 클래스 초기화 #50

donggeonoh committed:
feat: Chart view 추가 및 기본 레이아웃 설정, Mock data 추가 #50

donggeonoh committed:
feat: 코인 차트 뷰 Candlestick API 호출 및 데이터 추가 기능 구현 #50

donggeonoh committed:
Merge branch 'develop' into feat/50/implement-chart-layout

donggeonoh committed:
feat: 차트 그래프 네트워크 통신 구현 #50

donggeonoh committed:
feat: 그래프 API 데이터 추가 및 실시간 갱신 구현

HJEHA commented on BithumbYagomAcamedy/Controller/CoinChartViewController.swift:
```diff
@@ -0,0 +1,110 @@
+        coinChartView.leftAxis.enabled = false
+        coinChartView.doubleTapToZoomEnabled = false
+        
+//        coinChartView.xAxis.valueFormatter = createFormatter()
+//        coinChartView.delegate = self
```
주석이나 print() 구문은 제거 후 PR 올리시는게 좋을 것 같습니다:)

donggeonoh replied:
앗 죄송합니다 ㅠㅠ 빼고 올릴게요 ㅎㅎ

HJEHA commented on BithumbYagomAcamedy/Controller/CoinChartViewController.swift:
```diff
@@ -0,0 +1,110 @@
+//        coinChartView.delegate = self
+    }
+    
+    private func createFormatter() -> DefaultAxisValueFormatter {
+        return DefaultAxisValueFormatter { value, axis in
+            let dateFormatter = DateFormatter()
+            
+            dateFormatter.dateFormat = "MM-dd"
+            guard let interval = self.dataManager?.candlesticks[Int(value)].time else {
+                return ""
+            }
+            print(dateFormatter.string(from: Date(timeIntervalSince1970: TimeInterval(interval))))
+            print(interval)
+            return dateFormatter.string(from: Date(timeIntervalSince1970: interval))
+        }
+    }
```
Date 변환은 뷰컨이 아닌 데이터매니져나 다른 타입에서 하면 뷰컨의 역할이 줄어들 것 같습니다:)

donggeonoh replied:
이부분 현재 쓰이지 않는 부분이라서 지우도록 하겠습니다 ㅎㅎ

HJEHA commented on BithumbYagomAcamedy/Controller/CoinChartViewController.swift:
```diff
@@ -0,0 +1,110 @@
+
+extension CoinChartViewController: CoinChartDataManagerDelegate {
+    func coinChartDataManager(didSet candlesticks: [Candlestick]) {
+        let entries = candlesticks.enumerated().map { create(candlestick: $1, at: $0) }
+        let dataSet = create(dataSet: entries)
+        let data = CandleChartData(dataSet: dataSet)
```
여기서도 뷰컨에서 데이터를 가공하고 있는데 데이터매니저에서 가공을 완료한 후에 전달할 수는 없는건가요???:)

donggeonoh replied:
Charts 엔트리, 데이터셋을 모델에서 만들게된다면, 모델이 Charts 라이브러리를 import하는 상황이 되기 때문에
데이터 관리를 Model에서 하고 View에 쓰이는 Entry, DataSet Model은 컨트롤러에서 만들어주었습니다~
아무래도 저 작업을 VC에서 할건 아닌것같아 익스텐션으로 빼서 작업해보겠습니다.

HJEHA commented on BithumbYagomAcamedy/Network/HTTP/Model/APIModels/CandlestickAPI.swift:
```diff
@@ -0,0 +1,25 @@
+    
+    init(
+        orderCurrency: String = "BTC",
+        payment_currency: String = "KRW",
+        chart_intervals: String = "24h",
```
`payment_currency`, `chart_intervals`는 카멜 케이스를 사용하시는게 좋을 것 같습니다:)
chart_intervals의 경우 String 타입 보다는 열거형을 만드는 방법도 있을 것 같네요:)

donggeonoh replied:
앗 수정하겠습니다~

HJEHA reviewed (commented):
코멘트 남겼습니다 :)
확인 부탁드려요!

donggeonoh committed:
refactor: 코인 차트 VC의 DateFormatter 수정 등 PR 피드백 반영 #50

- 코인 차트 VC의 사용중이지 않는(createFormatter() 등) 메서드 삭제
- 코인 차트 VC에서 dateFormat을 수행하는 역할을 dataManger로 이동
- DateFormat enum 열거형 추가

HJEHA commented on BithumbYagomAcamedy/Model/CoinChartDataManager.swift:
```diff
@@ -0,0 +1,132 @@
+    }
+    private let symbol: String
+    private let formatType: DateFormat
+    var delegate: CoinChartDataManagerDelegate?
```
혹시 deinit 호출되는지 확인해보셨나요??

CoinChartDataManager랑 CoinChartViewController이 순환참조가 일어날 것 같습니다.
CoinChartDataManagerDelegate 프로토콜에 AnyObject를 상속해주면 class 타입만 채택이 가능하기 때문에 weak 키워드를 사용해서 순환참조를 방지할 수 있을 것 같습니다:)

donggeonoh replied:
헉! 큰 실수를 했네요..
확인해보니 deinit이 안불리고 있었어요... dataManger에서 weak로 들고있도록 수정하겠습니다.

donggeonoh committed:
feat: delegate 순환 참조 문제 해결

donggeonoh committed:
Merge branch 'develop' into feat/50/implement-chart-layout

donggeonoh committed:
style: CandlestickAPI 줄바꿈 띄어쓰기 수정
